### PR TITLE
[AutoTVM][Autoscheduler] Default build funcs inherit PassContext

### PR DESCRIPTION
**Pull request by AndrewZhaoLuo**
See https://github.com/apache/tvm/pull/11618 for old discussion

These funcs used building during the tuning process now inherit PassContext so we can inject things like pass instrumentation into the build process during tuning.

Note, due to the fact subprocesses are used for building by default, we can't naively do something like 

```
with PassContext(...):
   tune()
```
since new processes will not inherit the parents context.

But what does allow us to do is wrap the existing build functions with new PassContexts and call those instead making it a little more extensible.

### CODE REVIEW

**Comment by AndrewZhaoLuo** [comment]
Synced offline with @areusch, and created https://github.com/apache/tvm/issues/11656 to track this unintuitive behavior.

**Comment by areusch** [comment]
i'm good with this, but can you add a unit test to verify check_gpu is applied?

File: tests/python/integration/test_tuning.py [hunk 1]
```diff
@@ -180,6 +184,114 @@ def runner(target, dev):
...
+        results = []
+
+        tuner = RandomTuner(task)
+        tuner.tune(
+            n_trial=1,
+            measure_option=measure_option,
+            callbacks=(lambda _tuner, _inputs, rs: results.extend(rs),),
+        )
+
+        assert len(results) == 1
```
Comment (areusch) [inline tests/python/integration/test_tuning.py:290]: want to check the pass also succeeded? i think if one of those asserts fail we just get measure error here

Comment (areusch) [reply tests/python/integration/test_tuning.py:290]: ah ok--maybe the runner is where exceptions don't fail the test. good enough, if you've proven locally an exception causes the test to fail.

**Review by areusch — APPROVED** [review]

**Pull request merged**